### ISSUE TITLE
Some further usage examples 

### ISSUE THREAD

**Issue by nick-stephen** [mike-lischke/antlr4-c3#6]
Not a bug, a request. This project looks very hopeful for our usage, thanks.

To be really useful it would help to have some more concrete use cases on how to retrieve the incomplete symbol and its context, for example (from my experiments, where I'm not sure what to do.

I'm writing a simple filter grammar (imagine something grepping through logfiles) and want to auto-complete on something like:

fu = "incomplete string

When finding possible completions, I'd like to be able to have the context that I'm comparing a field called 'fu' (and thus be able to query all known values of 'fu') and that the token I'm trying to complete on is "incomplete string

Similarly we might not have the double-quotes on the incomplete string if we want the user to be able to type in a string without quotes...

An example or test case tying together the grammar with the symbol table and extracting the parse context + the half-complete symbol would be really helpful, I'm a little stuck here...

As an aside, if interested I've ported the completion core and tests to Java and may be able to release them back to this project. If so please context me.

**Comment by mike-lischke** [comment]
An incomplete string is a syntax errror and the walker might not find the path to the invocation point for candidate collection. But I haven't tested this scenario yet, as you usually don't have completion for strings (and literals in general). Have to think a bit how to get a context for that symbol.

Regarding the Java port: yes, sure, I'm interested in that. There is already a Kotlin port of the code (and I've written myself a C++ variant). I could publish all that together in this repo.

**Comment by nick-stephen** [comment]
Hi Mike,

Would it be possible for the walker to record the matching text and/or the set of tokens matching a given preferred rule, which would give the ability for the code trying to do completion to get not just the most recent token but the preferred rule’s full string? In the case I describe this would mean that we could have a rule called ‘partial_string_value’ matching an optional quote then a white-spaced string, and this would be a preferred rule which I could use in auto-complete and I’d be able to get the value. Today I can’t get the value since I can’t just use the last token, I need all the tokens that matched the ‘partial_string_value’ rule. I guess that information must be available during  the processRule evaluation?

I’ve started getting approvals to give back the java port, will keep you informed… I guess we could then move each port into separate sub-directories, I could prepare a cloned repo with a merge request once approvals are in place, which hopefully should be pretty quick.

Nick

On 14/08/2017, 18:01, "Mike Lischke" <notifications@github.com<mailto:notifications@github.com>> wrote:


An incomplete string is a syntax errror and the walker might not find the path to the invocation point for candidate collection. But I haven't tested this scenario yet, as you usually don't have completion for strings (and literals in general). Have to think a bit how to get a context for that symbol.

Regarding the Java port: yes, sure, I'm interested in that. There is already a Kotlin port of the code (and I've written myself a C++ variant). I could publish all that together in this repo.

**Comment by nick-stephen** [comment]
I believe I've found out how to extract the multi-token rule-string value as required to be able to do completions on things which span more than one token such as the given example.

It's done as an extension to collectCandidates() and a new field in the candidates object which maps a given candidate rule to its string fragment.

In my java implementation it looks like this:

~~~
        this.processRule(this.atn.ruleToStartState[startRule], 0, callStack, "\n");

        // now post-process the rule candidates and their last occurrences to find the rule strings
        for (int ruleId : candidates.rules.keySet()) {
            Map<Integer, Set<Integer>> shortcut = shortcutMap.get(ruleId);
            if (shortcut == null) {
                continue;
            }
            // select the right-most occurrence - the highest entry
            int startToken = Collections.max(shortcut.keySet());
            Set<Integer> endSet = shortcut.get(startToken);
            int endToken = endSet.isEmpty() ? 
                    caretTokenIndex - currentIndex : 
                    Collections.max(shortcut.get(startToken));
            // extract the matching text
            candidates.ruleStrings.put(ruleId, tokenStream.getText(tokenStream.LT(startToken), tokenStream.LT(endToken)));
        }
~~~

then for each candidate rule I have the last string in the parsed input which matched that rule.

HTH

Nick

**Comment by mike-lischke** [comment]
If I understand this right you want to extract the longest text that was matched when walking a candidate rule. What are you going to do with that text? It's not something you would show during code completion. Also, often candidate rules are not completely executed (since they are candidates, the caret position is somewhere in the text covered by those rules) and hence have no entry in the shortcutMap.

**Comment by mike-lischke** [comment]
Regarding the contribution of your Java code: glad to see the approval process has started. I'm still thinking about the repo structure and where to place the ports, since we must ensure we don't break the Node.js module setup. Maybe we just add a new folder called `ports` or `contributions` and exclude that in the package.json file?

**Comment by nick-stephen** [comment]
Hi Mike,

The idea is to extract the most recent (closest to the caret position) and longest text that matches when walking a candidate rule, so that completions of multi-token rules can have the partially-completed text… I’m still experimenting with this…  it’s not yet seeming to get all the string values I’d like, maybe your comment identifies the problem.

Here are a couple of use-cases for a technical-english grammar which performs filtering of log lines:

Imagine a log line like this:

Aug 19 00:07:00 info nicks-mbp cron[17522]: no path for address 0x10fab9000

And a filter like this:

    log text contains no path for

We might want to detect that the field being filtered on is the ‘log text’ field (which contains the full line) and the value to filter on is ‘no path for’ which in itself is a legal filter but which could be completed by scanning logs and finding that it could be completed to become ‘no path for address 0x10fab9000’ (or other addresses) based upon frequency of log messages.

To do this requires finding the ‘no path for’ text for the ‘value’ rule, and indeed also finding the ‘log text’ for the ‘field’ rule to give context for the search.

Similarly if typing:

  log t

I’d like to complete with ‘log text’ and ‘log timestamp’

And typing:

  log priority is

I’d like to propose completion with the values from the ‘priority’ field of log files, based upon their occurrence in the logs (e.g. ‘info’ in the above line).

Hope this is clear!

Concerning where to put in the source code base, I agree that separate subdirectories would be good, up to you if you want to create a ‘ports’ subdirectory, let me know…

Thx

Nick


On 19/08/2017, 11:50, "Mike Lischke" <notifications@github.com<mailto:notifications@github.com>> wrote:


If I understand this right you want to extract the longest text that was matched when walking a candidate rule. What are you going to do with that text? It's not something you would show during code completion. Also, often candidate rules are not completely executed (since they are candidates, the caret position is somewhere in the text covered by those rules) and hence have no entry in the shortcutMap.

**Comment by mike-lischke** [comment]
> Aug 19 00:07:00 info nicks-mbp cron[17522]: no path for address 0x10fab9000
>
> And a filter like this: log text contains no path for We might want to detect that the field being filtered on is the ‘log text’ field (which contains the full line) and the value to filter on is ‘no path for’ which in itself is a legal filter but which could be completed by scanning logs and finding that it could be completed to become ‘no path for address 0x10fab9000’ (or other addresses) based upon frequency of log messages.

This sounds to me like a special form of symbol lookup. Normally, in program languages, you collect available symbols while you parse the input (e.g. variable names) and when you invoke code completion and get a candidate that says: "here we can list var names" you take your collected variable names (maybe filtered by scope) and add them to the completion list. In your case it's the same: `address 0x10fab9000` or just `0x10fab9000` are the variables which you wanna show on code completion invocation. You don't need to get the text of the candidate rule (which is something completely different anyway).

What makes sense however is to scan backwards from the caret position and collect everything which has been typed already (up to a certain delimiter) and use that to filter the candidates list.

> Similarly if typing: log t I’d like to complete with ‘log text’ and ‘log timestamp’ 

This looks very much like completing keywords. Why do you want to scan your input first before offering `text` and `timestamp` as candidates, if you know these are valid values?

**Comment by nick-stephen** [comment]
Let's close the issue. Thanks!

**Comment by kingfolk** [comment]
Very appreciate about the work! I came across a problem and believe walker should have ability to go further even if there is syntax error before caret position. Suppose I have defined mysql grammar. And editing the following sentence,

```sql
select from |
```

where `'|'` represents caret position and it will get nothing by calling `collectCandidates`. Well, I can still do a little work to make it work. But It is also reasonable the walker keep walking if there is potential road to current request position. And I believe this is achievable by search algorithm like A-star. When walker walked to `select` and find no bridge to `from`, it search all neighbor bridges until it finally reach `from`. And this search should be constrained because it maybe very time-consuming.

Is it make sense? Thanks!
